### ISSUE TITLE
Add a procedural macro frontend

### ISSUE THREAD

**Issue by jplatte** [mozilla/uniffi-rs#1257]
This is a tracking issue for the proc-macro work, initially discussed in [#385](https://github.com/mozilla/uniffi-rs/issues/385#issuecomment-1098194211), on Matrix and recently in a video meeting.

### Outline
* Introduce an attribute `#[uniffi::export]` that can go onto functions and `impl` blocks
  * Attributes like `#[uniffi::default]` will be used to express binding properties that can't be expressed in Rust directly¹
* Introduce a derive attribute `uniffi::FfiConvertible` to allow a custom type to be used in the FFI API²
  * Attributes like `#[uniffi(default)]` will be used to express binding properties that can't be expressed in Rust directly³

### Progress
* [ ] Functions
  * [x] Argument passing
  * [x] Return value passing
  * [x] Fallible (returning `Result<T, E>`)
  * [ ] Argument defaults
* [ ] Types
  * [x] Primitives
  * [x] `Arc<_>`
  * [x] By-value structs and enums
    * [ ] Field defaults
  * [x] Enums
  * [ ] External types
* [ ] Parsing `#[doc = "…"]` attributes
* [x] `impl` blocks
  * [x] Associated functions (`static` methods in some target languages)
  * [x] (True) methods with `self` parameters
  * [x] Recognize `Self` keyword
  * [x] Constructors (`#[uniffi::constructor]`)
* [x] Traits
* [x] Callback interfaces

¹ Using custom syntax like `foo: u32 = 0` the parameter list would be possible, but it's undesireable because (a) it would be confusing and (b) it would degrade IDE functionality
² This could be handled by the same `#[uniffi::export]` attribute but that's undesireable for technical reasons
³ Syntax difference is warranted due to proc-macro peculiarities

I can elaborate on any of these footnotes if somebody wants to know more ^^

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/UNIFFI-168)
┆Issue Number: UNIFFI-168

**Comment by jplatte** [comment]
About the current status: Some things have their boxes ticked but don't have PRs yet. The reason for that is that most of the work I've been doing so far has not been under the assumption that we'll want to mix UDL and proc-macro interface definitions. However, recently I realized that this shouldn't be so hard and has a number of big advantages:

* Individual proc-macro bits can be tested and used ahead of achieving feature parity
  * This also means small PRs are possible, which makes collaboration on this much easier
* Incrementally porting an existing UDL-using codebase is possible

You can expect the first proc-macro PRs this week though 😉

**Comment by jplatte** [comment]
Some improvements happened in https://github.com/mozilla/uniffi-rs/pull/1282.

**Comment by jplatte** [comment]
Also I have a question that I wasn't sure whether to ask here or on Matrix:

Why does the current scaffolding contain a separate `FfiConverterTypeFooError` instead of implementing `RustBufferFfiConverter` / `FfiError` for the user's `FooError` type directly?

**Comment by bendk** [comment]
> Also I have a question that I wasn't sure whether to ask here or on Matrix:
> 
> Why does the current scaffolding contain a separate `FfiConverterTypeFooError` instead of implementing `RustBufferFfiConverter` / `FfiError` for the user's `FooError` type directly?

It's part of a general policy to not implement the trait on any user types, in case those types came from another crate.  It doesn't make that much sense for error types, but I guess there's a change.

Is this interfering with your work?  Maybe we should try to tackle #1250 to simplify things.

**Comment by jplatte** [comment]
Yeah, I think tackling the `FfiConverter` simplification first makes sense. I kinda wanted to do that after the proc-macro stuff, but that's now on a much slower path anyways, with the clean migration path figured out and lots of other important stuff being on my work todo list 😅

**Comment by jplatte** [comment]
An issue that I think hasn't been thought about with the current approach where the scaffolding code is generated by the macro directly is how type-specific scaffolding code (rather than function / method scaffolding code) is emitted. When I tried to move the only function using a `u16` parameter out of UDL, I got some undefined reference errors about u16 conversion code.

I can think of two possible solutions, but I don't know how feasible they are:

* Let every use of `#[uniffi::export]` emit the conversion code without any symbol prefix / suffix and use linker hints to let the linker de-duplicate the identical symbols
* Just add the conversion code to uniffi itself, there's a fixed number of builtin types that won't bloat resulting libraries that much and when statically linking the final library, unused ones might even be thrown out by the linker

**Comment by bendk** [comment]
That second bullet is how we handle builtin types currently.  I'm surprised that there would be an issue with a `u16` parameter, can you paste the error here?

**Comment by jplatte** [comment]
So it's only with `u16?` and I think it's related to `add_known_type` not recursing. I've tried a simple fix and it didn't seem to work though. I'll try to revisit it soon.

**Comment by bendk** [comment]
> So it's only with `u16?`

Haha I thought that's what *you* were saying.   I would expect any errors to apply to all numberic types, and possibly things like `String` as well.

In any case I think the fix should be simple, if it's still an issue when you revisit it, please paste the error message and I can take a look.

**Comment by jplatte** [comment]
I was trying to say that it only happens with `Option`s, not optional `u16`s specifically. Anyways, I'm pretty sure this PR will fix it: #1356

**Comment by jplatte** [comment]
This issue has definitely served its purpose ^^
